src/actions.rs

```diff
@@ -0,0 +1,886 @@
+pub struct NullActionChannel {
+    /// An identifier to distinguish this channel from others. Choose a meaningful string.
+    /// May be useful for debugging.
+    id: String,
+    /// The list of actions for this channel.
+    actions: Vec<NullAction>,
+}
+
+impl NullActionChannel {
+    /// Create a new NullActionChannel.
+    pub fn new(id: &str) -> Self {
+        Self {
+            id: id.to_string(),
+            actions: Vec::new(),
+        }
+    }
+
+    /// Add a pause action to this channel.
+    pub fn add_pause(&mut self, duration: Duration) {
+        self.add_action(NullAction::Pause {
+            duration: duration.as_millis() as u64,
+        });
+    }
+
+    /// Add the specified action to this channel.
+    pub fn add_action(&mut self, action: NullAction) {
+        self.actions.push(action);
+    }
+}
+
+/// A channel containing `Key` actions.
+#[derive(Debug, Clone)]
+pub struct KeyActionChannel {
+    /// An identifier to distinguish this channel from others. Choose a meaningful string.
+    /// May be useful for debugging.
+    id: String,
+    /// The list of actions for this channel.
+    actions: Vec<KeyAction>,
+}
+
+impl KeyActionChannel {
+    /// Create a new Key channel.
+    pub fn new(id: &str) -> Self {
+        Self {
+            id: id.to_string(),
+            actions: Vec::new(),
+        }
+    }
+
+    /// Add a pause action to this channel.
+    pub fn add_pause(&mut self, duration: Duration) {
+        self.add_action(KeyAction::Pause {
+            duration: duration.as_millis() as u64,
+        });
+    }
+
+    /// Add the specified action to this channel.
+    pub fn add_action(&mut self, action: KeyAction) {
+        self.actions.push(action);
+    }
+}
+
+/// A channel containing `Key` actions.
+#[derive(Debug, Clone)]
+pub struct PointerActionChannel {
+    /// An identifier to distinguish this channel from others. Choose a meaningful string.
+    /// May be useful for debugging.
+    id: String,
+    /// The pointer type. Can be `Mouse`, `Pen` or `Touch`.
+    pointer_type: PointerActionType,
+    /// The list of actions for this channel.
+    actions: Vec<PointerAction>,
+}
+
+impl PointerActionChannel {
+    /// Create a new Pointer channel.
+    pub fn new(id: &str, pointer_type: PointerActionType) -> Self {
+        Self {
+            id: id.to_string(),
+            pointer_type,
+            actions: Vec::new(),
+        }
+    }
+
+    /// Add a pause action to this channel.
+    pub fn add_pause(&mut self, duration: Duration) {
+        self.add_action(PointerAction::Pause {
+            duration: duration.as_millis() as u64,
+        });
+    }
+
+    /// Add the specified action to this channel.
+    pub fn add_action(&mut self, action: PointerAction) {
+        self.actions.push(action);
+    }
+}
+
+/// An ActionChannel is a sequence of actions of a specific type.
+/// When combined with other channels, you can think of them like a grid, with actions on the
```

Comment:
If you have the time, and ASCII diagram would be _amazing_ here!

Also, I think convention is to leave a blank line between the summary of a method/type (the first sentence) and what follows.

Reply:
I'll give it a go. The "channel" idea is just something I came up with to try to visualize it better. If you think a different term would fit better I'm open to it.

Reply:
Actually the spec has `input sources` so I could rename `Channel` to `InputSource` everywhere if that is preferred.

Reply:
The diagram looks great!

Ah, interesting. In that case, how about:
```rust
// Needed just so we don't expose a `webdriver` type in the public API
pub struct ActionSequence(WDActions::ActionSequence);

trait InputSource: Into<ActionSequence> {
    type Action;
    fn pause(&mut self, Duration) -> &mut Self;
    fn then(&mut self, Self::Action) -> &mut Self;
}

impl InputSource for NullActions { /* .. */ }
impl InputSource for KeyActions { /* .. */ }
impl InputSource for TouchActions {
    type Action = PointerAction;
    // ...
}
impl InputSource for MouseActions {
    type Action = PointerAction;
    // ...
}
impl InputSource for PenActions {
    type Action = PointerAction;
    // ...
}

pub struct Actions {
    sequences: Vec<ActionSequence>
}

impl Actions {
    fn and(&mut self, sequence: impl Into<ActionSequence>) -> &mut Self {
        self.sequences.push(sequence.into());
    }
}
```

Reply:
Very nice! Thanks!